src/core/ProjectJournal.cpp

```diff
@@ -51,84 +52,142 @@ ProjectJournal::ProjectJournal() :
-		JournallingObject *jo = m_joIDs[c.joID];
+		BatchCheckPoint backup;
 
-		if( jo )
+		// For every checkpoint (journaled object) in the last batch...
+		for (CheckPoint& restorePoint: restoreStack.back())
 		{
-			DataFile curState( DataFile::JournalData );
+			JournallingObject* jo = journallingObject(restorePoint.joID);
+			// TODO when can this be null?
+			if (!jo) { continue; }
+
+			// Create a backup of the object's current state
+			DataFile curState(DataFile::JournalData);
 			jo->saveState( curState, curState.content() );
-			m_undoCheckPoints.push( CheckPoint( c.joID, curState ) );
+			backup.push_back(CheckPoint(restorePoint.joID, curState));
 
+			// Restore object to its previous state
 			bool prev = isJournalling();
 			setJournalling( false );
-			jo->restoreState( c.data.content().firstChildElement() );
+			jo->restoreState(restorePoint.data.content().firstChildElement());
 			setJournalling( prev );
 			Engine::getSong()->setModified();
-			break;
 		}
+		restoreStack.pop_back();
+
+		// Keep trying until something is restored
+		if (backup.empty()) { continue; }
+
+		backupStack.push_back(backup);
+		return;
 	}
 }
 
+
+
+
 bool ProjectJournal::canUndo() const
 {
-	return !m_undoCheckPoints.isEmpty();
+	return !m_undoCheckPoints.empty();
 }
 
+
+
+
 bool ProjectJournal::canRedo() const
 {
-	return !m_redoCheckPoints.isEmpty();
+	return !m_redoCheckPoints.empty();
 }
 
 
 
+
 void ProjectJournal::addJournalCheckPoint( JournallingObject *jo )
 {
 	if( isJournalling() )
 	{
 		m_redoCheckPoints.clear();
 
+		// If we're not batching checkpoints, begin on a new one
+		if (m_batchingCount == 0 || m_undoCheckPoints.empty())
+		{
+			m_undoCheckPoints.emplace_back();
+		}
+		BatchCheckPoint& batch = m_undoCheckPoints.back();
+
+		// If this object already has a checkpoint in the batch, skip it
+		for (const CheckPoint& checkpoint: batch)
+		{
+			if (checkpoint.joID == jo->id()) { return; }
+		}
+
+		// Create a checkpoint and save it to the batch
 		DataFile dataFile( DataFile::JournalData );
 		jo->saveState( dataFile, dataFile.content() );
+		batch.push_back(CheckPoint(jo->id(), dataFile));
 
-		m_undoCheckPoints.push( CheckPoint( jo->id(), dataFile ) );
+		// Remove excessive checkpoints from the stack
 		if( m_undoCheckPoints.size() > MAX_UNDO_STATES )
 		{
-			m_undoCheckPoints.remove( 0, m_undoCheckPoints.size() - MAX_UNDO_STATES );
+			m_undoCheckPoints.erase(m_undoCheckPoints.begin(), m_undoCheckPoints.end() - MAX_UNDO_STATES);
 		}
 	}
 }
 
 
 
 
+/*! \brief Start adding checkpoints together as a batch
+ *
+ *  \param append - Don't create a new batch, use the last checkpoint
+ */
+void ProjectJournal::beginBatchCheckPoint(bool append)
+{
+	if (!isJournalling()) { return; }
+	// Begin on a new batch if we are not already batching, or if there is nothing to append to
+	if (append ? m_undoCheckPoints.empty() : m_batchingCount == 0) { m_undoCheckPoints.emplace_back(); }
+	++m_batchingCount;
```

Comment:
First, I found the variable name confusing. `batchingCount` sounds like "This is the number of CheckPoints that I am `batching`". Maybe subjective, though :grin:

Secondly, why is this variable changed if we add/remove checkpoints here, but not e.g. in `ProjectJournal::addJournalCheckPoint`? Intended?

Thirdly, I guess this variable would (if stuff like "secondly" would be changed) be equivalent to something like `m_undoCheckPoints.size()` or `m_undoCheckPoints.size() + m_redoCheckPoints.size()`?

Reply:
Ah! So the only reason for that variable is to make sure that `begin...` and `end...` match? Then it would all make sense.

An RAII version, like `BashPointGuard` would be cool (though, probably, less descriptive...).

Reply:
Yes, this was what I had in mind, too. It should not cause unused variable warnings, since `std::lock_guard` is [used the same way](https://en.cppreference.com/w/cpp/thread/lock_guard).

I'm OK with either begin/end or RAII, as long as it's clear what the counter is used for :grin: